gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -44,17 +46,81 @@
  * <p>Package-private for internal use.
  */
 class ChannelPool extends ManagedChannel {
-  private final ImmutableList<ManagedChannel> channels;
+  // refresh every 55 minutes
+  private static long refreshPeriod = 55 * 60;
+  // spread out the refresh between channels
+  private static long refreshDelay;
+  // maximum seconds to wait for old channel to terminate
+  private static long terminationWait = 60;
+
+  private final List<ManagedChannel> channels;
   private final AtomicInteger indexTicker = new AtomicInteger();
   private final String authority;
 
   /**
    * Initializes the channel pool. Assumes that all channels have the same authority.
    *
-   * @param channels a List of channels to pool.
+   * @param poolSize number of channels in the pool.
+   * @param channelFactory method to create the channels.
+   * @param executorService if set, schedule periodically refresh the channels.
    */
-  ChannelPool(List<ManagedChannel> channels) {
-    this.channels = ImmutableList.copyOf(channels);
+  ChannelPool(
+      int poolSize, final ChannelFactory channelFactory, ScheduledExecutorService executorService)
+      throws IOException {
+    this.channels = new ArrayList<>();
+    for (int i = 0; i < poolSize; i++) {
+      channels.add(channelFactory.createSingleChannel());
+    }
+
+    // if executerService is available, schedule channels to be refreshed every 55 minutes
+    if (executorService != null) {
+      // stagger the refresh to not overload GFE
+      refreshDelay = refreshPeriod / poolSize;
```

Comment:
I would highly recommend using a random jitter. Although you are staggering the refreshes in a single process, you can still get a thundering herd across multiple processes/machines.